File: wifi-handshake-capture/src/types/WiFiSniffer.ts
import type { TurboModule } from 'react-native/Libraries/TurboModule/RCTExport';
import { TurboModuleRegistry } from 'react-native';

export interface WiFiNetwork {
  bssid: string;
  ssid: string;
  signal: number;
  channel: number;
  security: string;
}

export type PacketType = 'EAPOL' | 'DEAUTH' | 'BEACON';

export interface HandshakePacket {
  timestamp: number;
  type: PacketType;
  bssid: string;
  clientMac: string;
  data: string;
  message?: 1 | 2 | 3 | 4;
}

export interface Spec extends TurboModule {
  scanNetworks(): Promise<WiFiNetwork[]>;
  startCapture(interfaceName: string): Promise<string>;
  stopCapture(): Promise<string>;
  sendDeauth(bssid: string, clientMac: string, count: number): Promise<boolean>;
  addListener(eventName: string): void;
  removeListeners(count: number): void;
}

export default TurboModuleRegistry.getEnforcing<Spec>('WiFiSniffer');

File: wifi-handshake-c/ios/WiFiSniffer/WiFiSniffer.m
#import <React/RCTBridgeModule.h>
#import <React/RCTEventEmitter.h>

@interface RCT_EXTERN_MODULE(WiFiSniffer, RCTEventEmitter)

RCT_EXTERN_METHOD(scanNetworks:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(startCapture:(NSString *)interfaceName
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(stopCapture:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(sendDeauth:(NSString *)bssid
                  clientMac:(NSString *)clientMac
                  count:(nonnull NSNumber *)count
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(addListener:(NSString *)eventName)
RCT_EXTERN_METHOD(removeListeners:(nonnull NSNumber *)count)

@end
